File: src/services/storage/tiered/CacheWarmer.ts
/**
 * Cache Warmer
 * Proactively populates hot tier cache on session start and access patterns
 */

import { getRedisObservationStore, type RedisObservationStore } from '../hot/RedisObservationStore.js';
import { getPostgresObservationStore, type PostgresObservationStore } from '../cold/PostgresObservationStore.js';
import { getPostgresProjectFactStore, type PostgresProjectFactStore } from '../cold/PostgresSummaryStore.js';
import { getRedisClient } from '../hot/RedisClient.js';
import { getTieredStorageConfig } from '../config.js';
import type { StoredObservation } from '../interfaces/IObservationStore.js';

export interface CacheWarmingStats {
  observationsWarmed: number;
  factsWarmed: number;
  durationMs: number;
}

export class CacheWarmer {
  private hotStore: RedisObservationStore | null = null;
  private coldStore: PostgresObservationStore;
  private factStore: PostgresProjectFactStore;
  private redisAvailable: boolean = false;
  private accessCounts: Map<number, number> = new Map();
  private accessPromotionThreshold: number = 3;

  constructor() {
    this.coldStore = getPostgresObservationStore();
    this.factStore = getPostgresProjectFactStore();
  }

  /**
   * Initialize the cache warmer
   */
  async initialize(): Promise<void> {
    try {
      const redisClient = getRedisClient();
      if (await redisClient.isAvailable()) {
        this.hotStore = getRedisObservationStore();
        this.redisAvailable = true;
      }
    } catch {
      this.redisAvailable = false;
    }
  }

  /**
   * Warm cache for a project on session start
   */
  async warmForProject(project: string): Promise<CacheWarmingStats> {
    const startTime = Date.now();
    const stats: CacheWarmingStats = {
      observationsWarmed: 0,
      factsWarmed: 0,
      durationMs: 0,
    };

    if (!this.redisAvailable || !this.hotStore) {
      stats.durationMs = Date.now() - startTime;
      return stats;
    }

    const config = getTieredStorageConfig();
    const hotTierCutoff = Date.now() - (config.redis.hotTierTTL * 1000);

    try {
      // Warm recent observations (last 48h)
      const recentObs = await this.coldStore.getRecent({
        project,
        since_epoch: hotTierCutoff,
        limit: 100,
      });

      for (const obs of recentObs) {
        try {
          await this.hotStore.cache(obs);
          stats.observationsWarmed++;
        } catch {
          // Ignore individual failures
        }
      }

      // Warm project facts (always hot - no TTL expiration for facts in hot tier)
      const facts = await this.factStore.getByProject(project, 50);
      for (const fact of facts) {
        try {
          const client = await getRedisClient().getClient();
          const key = `${getRedisClient().getKeyPrefix()}fact:${fact.id}`;
          await client.hSet(key, {
            id: fact.id.toString(),
            project: fact.project,
            fact_text: fact.fact_text,
            fact_type: fact.fact_type,
            confidence: fact.confidence.toString(),
          });
          // Facts don't expire - they're always relevant
          stats.factsWarmed++;
        } catch {
          // Ignore individual failures
        }
      }

      console.log(`[CacheWarmer] Warmed ${stats.observationsWarmed} observations, ${stats.factsWarmed} facts for ${project}`);
    } catch (err) {
      console.warn('[CacheWarmer] Error warming cache:', err);
    }

    stats.durationMs = Date.now() - startTime;
    return stats;
  }

  /**
   * Track access and promote frequently accessed observations
   */
  async onAccess(observationId: number): Promise<void> {
    if (!this.redisAvailable || !this.hotStore) {
      return;
    }

    // Increment access count
    const count = (this.accessCounts.get(observationId) || 0) + 1;
    this.accessCounts.set(observationId, count);

    // Check if should promote to hot tier
    if (count >= this.accessPromotionThreshold) {
      try {
        // Check if already in hot tier
        if (await this.hotStore.exists(observationId)) {
          return;
        }

        // Fetch from cold and cache
        const obs = await this.coldStore.getById(observationId);
        if (obs) {
          await this.hotStore.cache(obs);
          console.log(`[CacheWarmer] Promoted observation ${observationId} to hot tier after ${count} accesses`);
        }
      } catch {
        // Ignore promotion failures
      }
    }
  }

  /**
   * Batch warm multiple projects (e.g., for worktree support)
   */
  async warmForProjects(projects: string[]): Promise<CacheWarmingStats> {
    const combinedStats: CacheWarmingStats = {
      observationsWarmed: 0,
      factsWarmed: 0,
      durationMs: 0,
    };

    const startTime = Date.now();

    for (const project of projects) {
      const stats = await this.warmForProject(project);
      combinedStats.observationsWarmed += stats.observationsWarmed;
      combinedStats.factsWarmed += stats.factsWarmed;
    }

    combinedStats.durationMs = Date.now() - startTime;
    return combinedStats;
  }

  /**
   * Clear access tracking (e.g., on session end)
   */
  clearAccessTracking(): void {
    this.accessCounts.clear();
  }

  /**
   * Warm specific observations by IDs
   */
  async warmObservationsByIds(ids: number[]): Promise<number> {
    if (!this.redisAvailable || !this.hotStore || ids.length === 0) {
      return 0;
    }

    let warmed = 0;
    const observations = await this.coldStore.getByIds(ids);

    for (const obs of observations) {
      try {
        await this.hotStore.cache(obs);
        warmed++;
      } catch {
        // Ignore individual failures
      }
    }

    return warmed;
  }

  /**
   * Prefetch observations related to specific files
   */
  async warmForFiles(project: string, filePaths: string[]): Promise<number> {
    if (!this.redisAvailable || !this.hotStore || filePaths.length === 0) {
      return 0;
    }

    const config = getTieredStorageConfig();
    const hotTierCutoff = Date.now() - (config.redis.hotTierTTL * 1000);

    // Get observations that touched these files
    const observations = await this.coldStore.getRecent({
      project,
      files: filePaths,
      since_epoch: hotTierCutoff,
      limit: 50,
    });

    let warmed = 0;
    for (const obs of observations) {
      try {
        await this.hotStore.cache(obs);
        warmed++;
      } catch {
        // Ignore individual failures
      }
    }

    console.log(`[CacheWarmer] Warmed ${warmed} observations for files: ${filePaths.slice(0, 3).join(', ')}`);
    return warmed;
  }
}

// Singleton instance
let cacheWarmerInstance: CacheWarmer | null = null;

export function getCacheWarmer(): CacheWarmer {
  if (!cacheWarmerInstance) {
    cacheWarmerInstance = new CacheWarmer();
  }
  return cacheWarmerInstance;
}

export async function initializeCacheWarmer(): Promise<CacheWarmer> {
  const warmer = getCacheWarmer();
  await warmer.initialize();
  return warmer;
}

File: src/python/claude_mem/storage/tiered/cache_warmer.py
"""
Cache Warmer

Warms the hot tier cache on session start and promotes frequently accessed data.
"""

import logging
from datetime import datetime, timedelta
from typing import Optional

from claude_recall.models import QueryOptions
from claude_recall.storage.cold.observation_store import PostgresObservationStore
from claude_recall.storage.cold.summary_store import PostgresProjectFactStore
from claude_recall.storage.hot.observation_store import RedisObservationStore
from claude_recall.storage.cold.postgres_client import get_postgres_client
from claude_recall.storage.hot.redis_client import get_redis_client
from claude_recall.config import get_config

logger = logging.getLogger(__name__)


class CacheWarmer:
    """
    Warms the hot tier cache for optimal performance.

    Strategies:
    1. On session start: Load recent observations for the project
    2. On access: Promote frequently accessed data
    3. Background: Preload project facts
    """

    def __init__(self):
        self.postgres = get_postgres_client()
        self.redis = get_redis_client()

        self.cold_observations = PostgresObservationStore(self.postgres)
        self.cold_facts = PostgresProjectFactStore(self.postgres)
        self.hot_observations = RedisObservationStore(self.redis)

        self.config = get_config()
        self._access_counts: dict[int, int] = {}

    async def warm_for_project(self, project: str) -> int:
        """
        Warm cache for a specific project on session start.

        Loads recent observations (last 48 hours) into hot tier.
        Returns count of observations cached.
        """
        # Calculate time window
        now = datetime.now()
        hot_tier_hours = self.config.redis.hot_tier_ttl // 3600
        since = now - timedelta(hours=hot_tier_hours)
        since_epoch = int(since.timestamp() * 1000)

        # Fetch recent observations from cold tier
        options = QueryOptions(
            project=project,
            since_epoch=since_epoch,
            limit=500,  # Cap for performance
        )

        observations = await self.cold_observations.get_recent(options)
        logger.debug(f"Warming cache with {len(observations)} observations for {project}")

        # Cache in hot tier
        cached = 0
        for obs in observations:
            try:
                await self.hot_observations.cache(obs)
                cached += 1
            except Exception as e:
                logger.debug(f"Failed to cache observation {obs.id}: {e}")

        # Also load project facts (no TTL, always hot)
        await self._warm_project_facts(project)

        logger.info(f"Cache warmed for {project}: {cached} observations")
        return cached

    async def _warm_project_facts(self, project: str) -> int:
        """Load project facts into hot tier (no TTL)."""
        facts = await self.cold_facts.get_by_project(project)

        cached = 0
        for fact in facts:
            try:
                # Facts are stored differently - just track for now
                # Could add a RedisFacts store if needed
                cached += 1
            except Exception:
                pass

        return cached

    async def on_access(self, obs_id: int) -> bool:
        """
        Track access and potentially promote to hot tier.

        Returns True if promoted.
        """
        # Increment access count
        self._access_counts[obs_id] = self._access_counts.get(obs_id, 0) + 1
        count = self._access_counts[obs_id]

        # Promotion threshold
        threshold = 3

        if count >= threshold:
            # Check if already in hot tier
            if await self.hot_observations.exists(obs_id):
                return False

            # Fetch from cold tier and promote
            obs = await self.cold_observations.get_by_id(obs_id)
            if obs:
                try:
                    await self.hot_observations.cache(obs)
                    logger.debug(f"Promoted observation {obs_id} to hot tier")
                    return True
                except Exception:
                    pass

        return False

    async def warm_for_files(self, project: str, files: list[str]) -> int:
        """
        Warm cache for observations related to specific files.

        Useful when a user starts working on particular files.
        """
        from claude_recall.storage.cold.hybrid_search import PostgresObservationSearch

        search = PostgresObservationSearch(self.postgres)
        from claude_recall.models import HybridSearchOptions

        options = HybridSearchOptions(
            project=project,
            limit=50,
        )

        results = await search.search_by_files(files, options)

        cached = 0
        for result in results:
            obs = await self.cold_observations.get_by_id(result.id)
            if obs:
                try:
                    await self.hot_observations.cache(obs)
                    cached += 1
                except Exception:
                    pass

        logger.debug(f"Warmed cache with {cached} file-related observations")
        return cached

    async def warm_for_concepts(self, project: str, concepts: list[str]) -> int:
        """
        Warm cache for observations related to specific concepts.

        Useful when a user's prompt mentions specific topics.
        """
        from claude_recall.storage.cold.hybrid_search import PostgresObservationSearch

        search = PostgresObservationSearch(self.postgres)
        from claude_recall.models import HybridSearchOptions

        options = HybridSearchOptions(
            project=project,
            limit=50,
        )

        results = await search.search_by_concepts(concepts, options)

        cached = 0
        for result in results:
            obs = await self.cold_observations.get_by_id(result.id)
            if obs:
                try:
                    await self.hot_observations.cache(obs)
                    cached += 1
                except Exception:
                    pass

        logger.debug(f"Warmed cache with {cached} concept-related observations")
        return cached

    def clear_access_counts(self) -> None:
        """Clear access counts (e.g., at end of session)."""
        self._access_counts.clear()


# Singleton instance
_warmer: Optional[CacheWarmer] = None


def get_cache_warmer() -> CacheWarmer:
    """Get the global cache warmer."""
    global _warmer
    if _warmer is None:
        _warmer = CacheWarmer()
    return _warmer
